packages/@aws-cdk/aws-cloudformation/test/test.pipeline-actions.ts

```diff
@@ -0,0 +1,144 @@
+
+      const stackArn = cdk.ArnUtils.fromComponents({
+        service: 'cloudformation',
+        resource: 'stack',
+        resourceName: 'MyStack/*'
+      });
+      test.ok(_grantsPermission(statements, 'cloudformation:DescribeStacks', stackArn)
+              && _grantsPermission(statements, 'cloudformation:DescribeChangeSet', stackArn),
+              'The pipelineRole was given permissions to describe the stack & it\'s ChangeSets');
+      test.ok(_grantsPermission(statements, 'cloudformation:CreateChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to create the desired ChangeSet');
+      test.ok(_grantsPermission(statements, 'cloudformation:DeleteChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to delete the desired ChangeSet');
+
+      test.deepEqual(action.inputArtifacts, [artifact],
+                     'The inputArtifact was correctly registered');
+
+      test.ok(_hasAction(actions, 'AWS', 'CloudFormation', 'Deploy', {
+        ActionMode: 'CHANGE_SET_CREATE_REPLACE',
+        StackName: 'MyStack',
+        ChangeSetName: 'MyChangeSet'
+      }));
+
+      test.done();
+    }
+  },
+  ExecuteChangeSet: {
+    works(test: nodeunit.Test) {
+      const stack = new cdk.Stack();
+      const statements = new Array<PolicyStatementJson>();
+      const pipelineRole = {
+        addToPolicy: statement => { statements.push(statement.toJson()); }
+      } as iam.Role;
+      const actions = new Array<cpapi.Action>();
+      const stage = {
+        _attachAction: act => { actions.push(act); },
+        pipelineRole
+      } as cpapi.IStage;
+      new cloudformation.PipelineExecuteChangeSetAction(stack, 'Action', {
+        stage,
+        changeSetName: 'MyChangeSet',
+        stackName: 'MyStack',
+      });
+
+      const stackArn = cdk.ArnUtils.fromComponents({
+        service: 'cloudformation',
+        resource: 'stack',
+        resourceName: 'MyStack/*'
+      });
+      test.ok(_grantsPermission(statements, 'cloudformation:ExecuteChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to execute the desired ChangeSet');
+
+      test.ok(_hasAction(actions, 'AWS', 'CloudFormation', 'Deploy', {
+        ActionMode: 'CHANGE_SET_EXECUTE',
+        StackName: 'MyStack',
+        ChangeSetName: 'MyChangeSet'
+      }));
+
+      test.done();
+    }
+  }
+});
+
+interface PolicyStatementJson {
+  Effect: 'Allow' | 'Deny';
+  Action: string | string[];
+  Resource: string | string[];
+  Condition: any;
+}
+
+function _hasAction(actions: cpapi.Action[], owner: string, provider: string, category: string, configuration?: { [key: string]: any}) {
```

Comment:
I don't think you need the underscore here, making them not exported should effectively make them private.

Reply:
I've kinda got into the habit of making internal stuff names start with a `_` and I've gotten to like it. I'm okay to do without but I don't think it hurts... I found it makes it easier to navigate code by understanding from the function name sorting with a `_` that this is something Ic an't "just re-use" from some other module.

Reply:
You would love Go then :)